### PR TITLE
Enable data reader input layer fields

### PR DESCRIPTION
**Pull request by bvanessen**
Allow input layers to specify a specific data field from each data sample.  This PR changes the model from a monolithic input layer pulling in all fields into the data flow graph, to a mode where each key field is input via a distinct input layer.

### CODE REVIEW

File: src/layers/io/input_layer.cpp [hunk 1]
```diff
@@ -149,12 +130,15 @@ template <typename TensorDataType,
           El::Device Dev>
 std::vector<int> input_layer<TensorDataType, T_layout, Dev>::
 get_data_dims(DataReaderMetaData& dr_metadata, int child_index) const {
-  if(child_index == 0) {
+  if(child_index != 0) {LBANN_ERROR("get_data_dims: Invalid child index"); }
+  if(m_data_field.compare(INPUT_DATA_TYPE_SAMPLES) == 0) {
```
Comment (benson31) [inline src/layers/io/input_layer.cpp:134]:
```suggestion
  if (m_data_field == INPUT_DATA_TYPE_SAMPLES) {
```

File: src/layers/io/input_layer.cpp [hunk 2]
```diff
@@ -149,12 +130,15 @@ template <typename TensorDataType,
...
     return dr_metadata.data_dims[data_reader_target_mode::INPUT];
-  }else if(child_index == 1) {
-    return dr_metadata.data_dims[this->m_data_reader_mode];
+  }else if(m_data_field.compare(INPUT_DATA_TYPE_LABELS) == 0) {
+    return dr_metadata.data_dims[data_reader_target_mode::CLASSIFICATION];
+  }else if(m_data_field.compare( INPUT_DATA_TYPE_RESPONSES) == 0) {
+    return dr_metadata.data_dims[data_reader_target_mode::REGRESSION];
```
Comment (benson31) [inline src/layers/io/input_layer.cpp:139]:
```suggestion
  }
  else if (m_data_field == INPUT_DATA_TYPE_LABELS) {
    return dr_metadata.data_dims[data_reader_target_mode::CLASSIFICATION];
  }
  else if (m_data_field == INPUT_DATA_TYPE_RESPONSES) {
    return dr_metadata.data_dims[data_reader_target_mode::REGRESSION];
```

File: src/data_coordinator/buffered_data_coordinator.cpp [hunk 3]
```diff
@@ -290,21 +299,20 @@ bool buffered_data_coordinator<TensorDataType>::update_data_set(generic_data_rea
...
-      }
-    }
+  if(data_field.compare(INPUT_DATA_TYPE_SAMPLES) == 0) {
+    view_or_copy_tensor(*buf.m_input_buffers[input_data_type::SAMPLES], input_buffer);
+  }else if(data_field.compare(INPUT_DATA_TYPE_LABELS) == 0) {
+    view_or_copy_tensor(*buf.m_input_buffers[input_data_type::LABELS], input_buffer);
+  }else if(data_field.compare( INPUT_DATA_TYPE_RESPONSES) == 0) {
+    view_or_copy_tensor(*buf.m_input_buffers[input_data_type::RESPONSES], input_buffer);
+  }else {
+    LBANN_ERROR("Unknown data_field_type value provided: " + data_field);
```
Comment (benson31) [inline src/data_coordinator/buffered_data_coordinator.cpp:312]:
```suggestion
  if (data_field == INPUT_DATA_TYPE_SAMPLES) {
    view_or_copy_tensor(*buf.m_input_buffers[input_data_type::SAMPLES], input_buffer);
  }
  else if (data_field == INPUT_DATA_TYPE_LABELS) {
    view_or_copy_tensor(*buf.m_input_buffers[input_data_type::LABELS], input_buffer);
  }
  else if (data_field == INPUT_DATA_TYPE_RESPONSES) {
    view_or_copy_tensor(*buf.m_input_buffers[input_data_type::RESPONSES], input_buffer);
  }
  else {
    LBANN_ERROR("Unknown data_field_type value provided: " + data_field);
```

**Review by benson31 — COMMENTED** [review]
Just some initial comments after some light breakfast reading...

Maybe default the value of `data_field` to `"datum"` to ease compatibility?

File: src/data_readers/unit_test/data_reader_synthetic_test_public_api.cpp [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+  // Create a local copy of the RNG to check the synthetic data reader
+  lbann::fast_rng_gen ref_fast_generator;
+  ref_fast_generator.seed(lbann::hash_combine(42, 0));
+
+  // Initalize a per-trainer I/O thread pool
+  auto io_thread_pool = lbann::make_unique<lbann::thread_pool>();
+  io_thread_pool->launch_pinned_threads(1, 1);
+
+  std::set<std::string> active_data_fields = {"samples"};
+  active_data_fields.insert(GENERATE("labels", "responses"));
```
Comment (benson31) [inline src/data_readers/unit_test/data_reader_synthetic_test_public_api.cpp:77]:
```suggestion
  active_data_fields.insert(GENERATE(std::string("labels"),
                                     std::string("responses")));
```
Apple clang was whining about this one...

**Review by benson31 — APPROVED** [review]
Overall this looks fine. I'd say we should run clang-format on the diff:
```
git diff -U0 --no-color develop HEAD -- "*.cpp" "*.hpp" | /clang/prefix/share/clang/clang-format-diff.py -i -p1
```
(or do a `find` to find that python file)

Also, Apple Clang whines that:
```
[451/556] Building CXX object unit_test/CMakeFiles/mpi-catch-tests.dir/__/src/data_readers/unit_test/data_reader_smiles_fetch_datum_test.cpp.o
../src/data_readers/unit_test/data_reader_smiles_fetch_datum_test.cpp:193:13: warning: unused variable 'Max_seq_len' [-Wunused-variable]
  const int Max_seq_len = 56;
            ^
1 warning generated.
```
I wasn't going to complain, but since you have now touched that file (copyright change), we should clean that.

**Comment by bvanessen** [comment]
@timmoon10 Bamboo now matches develop

File: src/proto/layers.proto [hunk 5]
```diff
@@ -415,7 +415,7 @@ message Layer {
   // Input layers //
   //////////////////
   message Input {
-    string target_mode = 3;       // Options: "classification", "regression", "reconstruction", "label_reconstruction", "N/A" (default)
+    string data_field = 1; // legacy names are: datum, label, regression
```
Comment (timmoon10) [inline src/proto/layers.proto:418]:
```suggestion
    string data_field = 1; // legacy names are: samples, labels, responses
```

**Review by timmoon10 — APPROVED** [review]
Overall looks good. It's a pretty big API change with the input layers, and possibly a painful one, but I think this is the right design.

The next step will be extending this generalized API into the data readers, so we can have arbitrarily named data fields.

File: include/lbann/data_coordinator/buffered_data_coordinator.hpp [hunk 6]
```diff
@@ -63,6 +63,19 @@ class buffered_data_coordinator : public data_coordinator {
...
+    //   for (const auto& buf_map : m_data_buffers) {
+    //     const data_buffer_map_t& buffer_map = buf_map;
+    //     for (auto& [mode, buffer] : buffer_map) {
+    //       buffer->initialize_buffer_for_data_field(data_field, m_comm);
+    //     }
+    //   }
+    // }
+
+    // setup_data_fields(get_trainer().get_max_mini_batch_size());
+
```
Comment (benson31) [inline include/lbann/data_coordinator/buffered_data_coordinator.hpp:78]: Remove dead code?
```suggestion
```

**Review by benson31 — APPROVED** [review]

**Pull request merged**